### PR TITLE
feat(auth): add confirmSignUp API

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes
PR adds `confirmSignUp` API. This API removes the default true value for ForceToAliasCreation in v5 and passes it as serviceOptions to match behavior with the underlying service

**Pending:**
- add userContextData

#### Description of how you validated changes

I created a sample app where I tested a successful signUP flow: signUp -> confirmSignUp and confirmed that a user was created and confirmed in the user pool.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/apis/confirmSignUp.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	ConfirmSignUpRequest,
+} from '../../../types';
+import { CustomAttribute, CognitoConfirmSignUpOptions } from '../types';
+import { assertValidationError } from '../../../errors/utils/assertValidationError';
+import { AuthValidationErrorCode } from '../../../errors/types/validation';
+import { ConfirmSignUpException } from '../types/errors';
+import { confirmSignUpClient } from '../utils/clients/ConfirmSignUpClient';
+
+/**
+ * Confirms a user account
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/apis/confirmSignUp.ts:18]:
```suggestion
 * Confirms a new user account.
```

File: packages/auth/src/providers/cognito/utils/clients/ConfirmSignUpClient.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+export async function confirmSignUpClient(
+	params: ConfirmSignUpClientInput
+): Promise<ConfirmSignUpCommandOutput> {
+	const client = new UserPoolHttpClient(UserPoolClient.region);
+	const result: ConfirmSignUpCommandOutput =
+		await client.send<ConfirmSignUpCommandOutput>('ConfirmSignUp', {
+			...params,
+			ClientId: UserPoolClient.clientId,
+		});
+	return result;
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/utils/clients/ConfirmSignUpClient.ts:29]: Nit, can we shorten this a little?

```suggestion
	return client.send<ConfirmSignUpCommandOutput>('ConfirmSignUp', {
			...params,
			ClientId: UserPoolClient.clientId,
    });
```

File: packages/auth/src/types/requests.ts [hunk 3]
```diff
@@ -58,3 +58,21 @@ export type SignUpRequest<
 	password: string;
 	options?: AuthSignUpOptions<UserAttributeKey, ServiceOptions>;
 };
+
+
+/**
+ *  Constructs a `confirmSignUp` request.
+ *
+ * @param username - a standard username, potentially an email/phone number
+ * @param confirmationCode- the user's confirmation code sent to email or cellphone
```
Comment (jimblanc) [inline packages/auth/src/types/requests.ts:67]:
```suggestion
 * @param confirmationCode - the user's confirmation code sent to email or cellphone
```

**Review by jimblanc — APPROVED** [review]
LGTM, some minor nits

File: packages/auth/src/providers/cognito/apis/confirmSignUp.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		AuthValidationErrorCode.EmptyConfirmSignUpCode
+	);
+
+	await confirmSignUpClient({
+		Username: username,
+		ConfirmationCode: confirmationCode,
+		ClientMetadata:
+			options?.serviceOptions?.clientMetadata ?? config.clientMetadata,
+		ForceAliasCreation: options?.serviceOptions?.forceAliasCreation,
+		// TODO: handle UserContextData
```
Comment (Samaritan1011001) [inline packages/auth/src/providers/cognito/apis/confirmSignUp.ts:53]: Is this dependent on something else to be finished?

File: packages/auth/src/providers/cognito/apis/confirmSignUp.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	assertValidationError(
+		!!username,
+		AuthValidationErrorCode.EmptyConfirmSignUpUsername
+	);
+	assertValidationError(
+		!!confirmationCode,
+		AuthValidationErrorCode.EmptyConfirmSignUpCode
+	);
+
+	await confirmSignUpClient({
```
Comment (Samaritan1011001) [inline packages/auth/src/providers/cognito/apis/confirmSignUp.ts:47]: Not sure if we plan on supporting this but what about the AnalyticsMetadata? 
https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmSignUp.html#CognitoUserPools-ConfirmSignUp-request-AnalyticsMetadata

File: packages/auth/src/providers/cognito/utils/clients/ConfirmSignUpClient.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	| 'ForceAliasCreation'
+	| 'UserContextData'
+>;
+
+export async function confirmSignUpClient(
+	params: ConfirmSignUpClientInput
+): Promise<ConfirmSignUpCommandOutput> {
+	const client = new UserPoolHttpClient(UserPoolClient.region);
+	const result: ConfirmSignUpCommandOutput =
+		await client.send<ConfirmSignUpCommandOutput>('ConfirmSignUp', {
```
Comment (Samaritan1011001) [inline packages/auth/src/providers/cognito/utils/clients/ConfirmSignUpClient.ts:25]: I know we haven't been doing it in the other APIs as well but should we be swallowing and retry for some of the exceptions and only rethrow ones that makes sense for customers to know about? 

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmSignUp.html#CognitoUserPools-ConfirmSignUp-request-AnalyticsMetadata:~:text=empty%20HTTP%20body.-,Errors,-For%20information%20about

Comment (israx) [reply packages/auth/src/providers/cognito/utils/clients/ConfirmSignUpClient.ts]: The good thing is that `confirmSignUp` is retriable if a there is a code mismatch. So not something that customers will be blocked on confirming their accounts.

File: packages/auth/src/providers/cognito/apis/confirmSignUp.ts (hunk 4 above)
Comment (israx) [reply packages/auth/src/providers/cognito/apis/confirmSignUp.ts:53]: yes good question. That depends on a current v5 implementation that is to support the advance security feature in Cognito. There is a plan already to refactor and migrate it into v6.

**Review by jimblanc — APPROVED** [review]

**Review by Samaritan1011001 — APPROVED** [review]

**Pull request merged**